documentation/proposals/Proposal - Better Strings.md

```diff
@@ -0,0 +1,102 @@
+# Summary
+A rich, more well-defined native string API.
+
+- Written against Silk.NET **1.0.2**.
+- Slated for Silk.NET **2.0.0**.
+
+# Contributors
+- Dylan P, Ultz Limited
+
+# Current Status
+- [x] Proposed
+- [ ] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decisions
+Currently, the representation of native strings in Silk.NET (and C# as a whole) is an utter mess.
+At the moment we're using `char*` everywhere which isn't technically correct, as a number of users
+have pointed out. This discussion aims to create a richer set of native string APIs to combine
+technical correctness, ease of marshalling, and ease of use.
+
+This is in the form of Char8, Char16, and Char32 which represent the various lengths of characters
+in various encodings. These CharXXs can be combined into a StringXX. All types can be implicitly
+converted to and from their existing C#/.NET types, so that the user doesn't have to deal with much
+
+# Proposed API
+Proposed Silk.NET.Core additions:
+```cs
+public struct Char8
+{
+    private byte _value; // 8-bit character
+    public Char8(byte v) => _value = v;
+    public static implicit operator Char8(char c);
+    public static implicit operator Char8(byte c);
+    // and vice versa
+    public char ToChar();
+    public byte ToByte();
+    public uint ToUInt32();
+}
+
+public struct Char16
+{
+    private char _value; // 16-bit character
+    public Char16(char v) => _value = v;
+    // insert appropriate operators and methods here as seen in Char8
+}
+
+public struct Char32
+{
+    private uint _value; // 32-bit character
+    public Char32(char v0, char v1) => _value = ((uint)v0 << 16) + (uint)v1;
+    // insert appropriate operators and methods here as seen in Char8
+}
+
+public ref struct String8
+{
+    private Span<Char8> _value;
+    public String8(Span<Char8> v) => _value = v;
+    public static implicit operator String8(Span<Char8> v);
+    public static implicit operator String8(string x); // not recommended for use, as we'll have to assume the encoding is UTF8.
+    public static implicit operator String8(Char8* v); // we don't know the length here, so the backing span will just have a length of 1
```

Comment:
This isn't how most C APIs work. A C `char* `or `wchar*` without an associated length is null terminated, so the proper thing to do would be:
```csharp
var s = new Span<Char8>(v, int.MaxValue);
return s.Slice(0, s.IndexOf((byte)'\0')).AsString()
```

Reply:
You could manually track the backing pointer + length so this doesn't need to be a ref struct. Then you could use `length == -1` to indicate "null-terminated, length unknown"
If you don't want to incur this as an up front cost.